### ISSUE TITLE
Explain commit hooks

### ISSUE THREAD

**Issue by warpech** [Starcounter/starcounter.io#35]

**Comment by warpech** [comment]
Example of using commit hooks: https://github.com/StarcounterSamples/UserAdmin/blob/master/src/UserAdmin/Api/CommitHooks.cs
 (lines 10-20)

**Comment by dan31** [comment]
Wrote an article regarding commit hooks:
http://starcounter.io/guides/transactions/commit-hooks/
@per-samuelsson @miyconst please review when possible.

**Comment by miyconst** [comment]
The article is good and the content is right. Some tips to improve:

- Add left side menu and right side content block like here: http://starcounter.io/guides/database/.
- Every sentence on the page is very long, would be good to shorten them.
- Add a link to explain transaction re-trying mechanism. 
- Add a note that detaching hooks is not possible right now.
- Add a note that accessing json view-model from a hook is not possible.

**Comment by per-samuelsson** [comment]
Nice one, @dan31!

Two short comments:

> Currently, the first argument passed by Starcounter to an event handler is always `null` [...]

I'd rather **change** that to say that the first argument is an implementation detail, and should not normally be used. And maybe mention it could be `null`.

And you could also **add** to the last section that CRUD operations introduced *inside* a hook is *not* triggering additional hooks.

And comments on what @miyconst commented:

> Every sentence on the page is very long, would be good to shorten them.

Agree.

> Add a link to explain transaction re-trying mechanism.

Don't think that's necessary, although it wouldn't be completely off either. They are just not that related IMO.

**Comment by miyconst** [comment]
>Don't think that's necessary, although it wouldn't be completely off either. They are just not that related IMO.

Just make those words be a link to another page, would be good for navigation.

**Comment by per-samuelsson** [comment]
On a related note, am I the only one feeling that event-based approach we choose when designing the API feels a bit outdated? 

I know, we've where through things over and over, and on top of that, it's really not the time to start making changes I think (considering everything in the pipe). But I would really favor an approach promoting lambdas, hence this

```cs
Hook<Person>.CommitInsert(p => p.FillBlanks());
```

rather than this

```cs
Hook<Person>.CommitInsert += (sender, p) => { p.FillBlanks(); }
```

That would make deregistration (which we don't yet support) a little more verbose maybe, but I think thats a marginal case.

**Comment by per-samuelsson** [comment]
> That would make deregistration (which we don't yet support) [...]

Hmm, reviewing the code now, it looks like I might actually have implemented that but forgot I did. I'll review that too, so don't do this until I give feedback:

> Add a note that detaching hooks is not possible right now.

**Comment by per-samuelsson** [comment]
> so don't do this until I give feedback

Nah, false alarm. It raise an exception runtime if you try, I can see that now. So adding that section is in place, as @miyconst point out.

**Comment by per-samuelsson** [comment]
More feedback btw:

1. Rename "TestCH" to like "TestHooks". Those kind of abbrevations feel to much C/C++ IMO. :smile:
2. In all `Console.Writeline` output, remove cryptic "CH" and "TR" prefixes. The sample is simple enough to understand without them.

**Comment by dan31** [comment]
@miyconst, could you come up with a particular example for this: "accessing json view-model from a hook is not possible", when is this necessary? better with some code snippets.

**Comment by miyconst** [comment]
Well, this is not something required, but something developers may try to.

Now rethinking this issue with the latest knowledge of Starcounter, I narrowed the scope. It happens because the hook and the view-model are in different transactions/scopes.

An example:

#### IndexPage.cs

```cs
partial class IndexPage : Page {
    protected override void OnData() {
        base.OnData();

        Hook<Value>.CommitInsert += this.OnValueInserted;
    }

    protected void OnValueInserted(object sender, Value entity) {
        this.Value.Data = entity;
        this.Session.CalculatePatchAndPushOnWebSocket();
    }
}
```

#### Program.cs

```cs
[Database]
public class Value {
    public DateTime Date { get; set; }
}

class Program {
    static void Main() {
        Handle.GET("/index", () => {
            IndexPage page = new IndexPage() {
                Data = null,
                Session = new Session(SessionOptions.PatchVersioning)
            };

            return page;
        });

        Handle.GET("/insert", () => {
            Db.Transact(() => {
                new Value() {
                    Date = DateTime.Now
                };
            });

            return "New object inserted";
        });
    }
}
```

This works perfect. As soon as `/insert` called, the `/index` page gets updated. Now, if I wrap the `/index` handler into a `Db.Scope` the things stop working.

```
Starcounter.DbException: ScErrTransactionLockedOnThread (SCERR4031): Can't modify the state of the transaction.
```

**Comment by dan31** [comment]
Reflected most of remarks except one regarding @miyconst view-model comment. Also got request from @alemoi and @careri to better explain a moment when commit hook is triggered. I'm going to select a right example for it with commit rollbacks.

**Comment by careri** [comment]
@miyconst as far as I've been able to figure out you need to register hooks at the main program thread during startup. right? This goes for creating indices as well?

An other opinion about the CommitHooks is that they are all actually pre commit hooks. They are triggered before the commit is actually written as far as I can tell. You have no idea if the commit that triggered the hook will be rollbacked after due to a conflict.
According to the documentation the events for insert or update are called after the transaction is fully written, but is this really true? 
http://starcounter.io/guides/transactions/commit-hooks/

**Comment by miyconst** [comment]
@careri most of the time I use hooks in static constructor of a class. I am not sure whether it is main program thread.
 An example:

```cs
[Database]
public class Order {
    public DateTime Date;
    public QueryResultRow<OrderItem> Items;

    static Order() {
        Hook<Order>.CommitInsert += (sender, entity) => { entity.Date = DateTime.Now; }
        Hook<Order>.BeforeDelete += (sender, entity) => { entity.Items.ForEach(x => x.Delete()); }
    }
}
```

We also use hooks to notify applications about user sign in state. SignIn application creates new SystemUserSession, and dependent applications `Hook<SystemUserSession>.CommitInsert` and update their state according to currently signed in user.

I know that a `Db.Transact` can be called many times in case of transaction conflict, but how a transaction might call hooks but end up not being commited? I think only exception can stop a transaction from being commited. In this case application won't work anyway, and executed hooks does not really matter.

**Comment by per-samuelsson** [comment]
> @miyconst as far as I've been able to figure out you need to register hooks at the main program thread during startup. right? 

> No. You can register them whenever.

> This goes for creating indices as well?

Not sure about that one. I don't think so, if I should guess. But its not related to this thread. Can you open a question issue about that one, so we can get a correct answer?

> An other opinion about the CommitHooks is that they are all actually pre commit hooks. They are triggered before the commit is actually written as far as I can tell. You have no idea if the commit that triggered the hook will be rollbacked after due to a conflict.

They are actually not "pre commit" either - they are *part of* a comitting transaction. They kind of need to obey to what happens to the transaction overall, and hence they can not influence "other" functionality, like if the commit is to fail, etc. The important design criteria is: **whatever stuff you do in your hook will meet the same destiny as fellow CRUD's in the same transaction**. If those finally get commited, so will yours. If they don't, neither will yours.

Pseudo-algoritm
```
INSERT
UPDATE
DELETE
(over and over)
void HostExecCommit() {
  RunHooksSoTheyCanAddToOrFilterOrModifyTransactionData();
  WriteEverythingToSecuredSpace();
}
```

> According to the documentation the events for insert or update are called after the transaction is fully written, but is this really true? 
http://starcounter.io/guides/transactions/commit-hooks/

See above.

**Comment by per-samuelsson** [comment]
The mindset to get into as a hook developer: "What do I want to add (insert), change (update) and remove (delete) if the current transaction should succeed".

**Comment by careri** [comment]
Yes, pre was the wrong name for it. I know how it works, but couldn't find the correct label.
But this example put's the finger on the problem.
https://github.com/StarcounterSamples/UserAdmin/blob/master/src/UserAdmin/Api/CommitHooks.cs

The pages are changed even though the commit might rollback. In most cases I would benefit from having PostCommit-hooks, meaning hooks that are actually called after the transaction is written and there are no changes of rollback. The same goes for your example code.

I can see cases where the current commit hooks have a purpose for triggering data changes that shall only be written if the commit actually succeeds.
But I would rather have hooks that are called after the data is written.

**Comment by miyconst** [comment]
>The pages are changed even though the commit might rollback.

The commit might rollback only if an exception occurs. Otherwise it will commit after one or more tries.

It's not an issue if the page gets updated on each try, and the app is not usable if commit got aborted because of an exception.

So this example works perfectly fine.

**Comment by careri** [comment]
Yes of course it works. But there is now need to trigger the hook in your example until the change is actually written.

The same goes for IndexPage.cs above, the Data-property is bound to an instance that might actually not exist if e.g. a UniqueConstraintViolation is thrown after the hook is called

**Comment by miyconst** [comment]
@careri I agree that this cases could/should be replaced with post-commit hooks, but if we have to select only pre-commit or post-commit hooks, then I prefer pre-commit one.

The pre-commit hooks work fine for necessary data changes and UI updates. The post-commit hooks would work only for UI updates.

**Comment by careri** [comment]
Well that's only true for the BeforeDelete-event, the insert or update events would work just fine with the post event hook. I usually want to react to data changes once I'm sure they are really written. Like sending messages to clients (not UI).

With the pre/during-commit hook I have to write some informational item for the data change that I then have to query for in another transaction to check if the item is really written. I would rather have only post-commit

**Comment by per-samuelsson** [comment]
> With the pre/during-commit hook I have to write some informational item for the data change that I then have to query for in another transaction to check if the item is really written. I would rather have only post-commit.

We could certainly consider extending Starcounter to support that too, but it seems to me its more of a marginal case. As a general feature, if I'd had to guess what was most useful between the two, I'd say the one we have today. That's just a guess though.

The thing with the second alternative - post-commit hooks - is that it kind of suffers from the same thing a feature such as `IsNew()` previously exposed. The instance in time you take an action on that condition, the condition might have changed. Yes, you could probably make up for that by subscribing to all mod ops (CUD), but that would still be kind of advanced.

>  In most cases I would benefit from having PostCommit-hooks, meaning hooks that are actually called after the transaction is written and there are no changes of rollback.

If they are rolled back, or deleted/changed by some transaction in between you get noticed about them - that's kind of the same thing, isn't it? You just can't rely on any type of hooks / logic like that if you want to produce something that is 100% accurate.

**Comment by per-samuelsson** [comment]
As a bottomline for this particular issue, I'd say we should just make sure we *accurately document how things work today*. It's quite unlikely we'll change the current behavior in any near future, and we must try separate what is requests for new features, and what it is we can improve right now - i.e. document what we have.

@careri, if you feel strongly about post-commit hooks, and that they should really make life easier for you, I'm happy to open up a new feature request issue for that, and we can move the discussion there. We could try together to describe the simplest possible case that could really benefit from that kind of feature, and that we have no idea on how to make it work by other (realistic) means. And have that as the catalyst for any future decision / planning around such a new feature. Just let me know how important you think it is and if you wish, and I'll spawn up a new issue for it.

**Comment by careri** [comment]
Well the first time I realized the problem when I was writing the word search engine in the old SCDB (1.5). 
1. A user changed the name of a product. 
2. A commit hook reacted to the name change and triggered an update och the word cache (not persistent)

So far so good, but then we started to get customer issues about not getting matches. Sometimes the commits was rolledback due to different constraints. The cache was already updated with the non committed data.

So the final solution was that during the commit-hook I create a persistent work item with the object ID of the changed object. Then a worker thread reads the work items and applies them to the word cache.


Another example is the message queues I'm implementing now, each time an item is available to a queue I fire an event that it shall try to distribute the item. But since the items have a unique constraint on a SortID I sometimes get conflicts. So I end up firing of events for items that doesn't exist. I solved this by starting a timer that tries to reread the added item after a specified delay.

For internal changes that are to be written in the same transaction the current implementation of Hooks is great. 

Another example is aggregations that I've got in my application. If one of the items in the aggregations is changed I refresh the aggregation. This is also triggered by commit hooks, so I might end up aggregating NOOP if the item that triggered the commit hook was rolledback.

Maybe I'm doing it the wrong way, but I would really like to see the post commit hooks/triggers. I', fine with the object not being in the exact same state as triggered the hook.
